Firestore/core/include/firebase/firestore/timestamp.h

```diff
@@ -0,0 +1,173 @@
+ * resolution in UTC Epoch time. It is encoded using the Proleptic Gregorian
+ * Calendar which extends the Gregorian calendar backwards to year one. It is
+ * encoded assuming all minutes are 60 seconds long, i.e. leap seconds are
+ * "smeared" so that no leap second table is needed for interpretation. Range is
+ * from 0001-01-01T00:00:00Z to 9999-12-31T23:59:59.999999999Z.
+ *
+ * @see
+ * https://github.com/google/protobuf/blob/master/src/google/protobuf/timestamp.proto
+ */
+class Timestamp {
+ public:
+  /**
+   * Creates a new timestamp representing the epoch (with seconds and
+   * nanoseconds set to 0).
+   */
+  Timestamp();
+
+  /**
+   * Creates a new timestamp.
+   *
+   * @param seconds The number of seconds of UTC time since Unix epoch
+   *     1970-01-01T00:00:00Z. Must be from 0001-01-01T00:00:00Z to
+   *     9999-12-31T23:59:59Z inclusive.
+   * @param nanoseconds The non-negative fractions of a second at nanosecond
+   *     resolution. Negative second values with fractions must still have
+   *     non-negative nanoseconds values that count forward in time. Must be
+   *     from 0 to 999,999,999 inclusive.
+   */
+  Timestamp(std::int64_t seconds, std::int32_t nanoseconds);
+
+  /**
+   * Creates a new timestamp with the current date.
+   *
+   * If <chrono> is available, precision will correspond to the precision of
+   * `std::chrono::system_clock`. Otherwise, the timestamp will have second
+   * precision.
+   *
+   * @return a new timestamp representing the current date.
+   */
+  static Timestamp Now();
+
+  /**
+   * The number of seconds of UTC time since Unix epoch 1970-01-01T00:00:00Z.
+   */
+  std::int64_t seconds() const {
+    return seconds_;
+  }
+
+  /**
+   * The non-negative fractions of a second at nanosecond resolution. Negative
+   * second values with fractions still have non-negative nanoseconds values
+   * that count forward in time.
+   */
+  std::int32_t nanoseconds() const {
+    return nanoseconds_;
+  }
+
+  /**
+   * Converts `std::time_t` to a Timestamp.
+   *
+   * @param seconds_since_unix_epoch The number of seconds of UTC time since
+   *     Unix epoch 1970-01-01T00:00:00Z. Can be negative to represent dates
+   *     before the epoch.
+   *
+   *     Note that while the epoch of `std::time_t` is unspecified, it's usually
+   *     Unix epoch. If this assumption is broken, this function will produce
+   *     incorrect results.
+   *
+   * @return a new timestamp with the given number of seconds and zero
+   *     nanoseconds.
+   */
+  static Timestamp FromTimeT(std::time_t seconds_since_unix_epoch);
+
+#if !defined(_STLPORT_VERSION)
+  /**
+   * Converts `std::chrono::time_point` to a Timestamp.
+   *
+   * @param time_point The time point with system clock's epoch, which is
+   *     presumed to be Unix epoch 1970-01-01T00:00:00Z. Can be negative to
+   *     represent dates before the epoch.
+   *
+   *     Note that while the epoch of `std::chrono::system_clock` is
```

Comment:
This assumption (and the mirroring one in `ToTimeT`) is unfortunate, but without making it, I don't think it's possible to provide any conversions at all.

Reply:
It seems to be pretty safe assumption. Anyone who's violating it is probably already very aware that they can't rely on most time conversion functions.